File: java_maven/src/main/java/edu/usc/ict/superglu/core/Message.java
package edu.usc.ict.superglu.core;

import edu.usc.ict.superglu.ontology.mappings.MessageTemplate;
import edu.usc.ict.superglu.util.SerializationConvenience;
import edu.usc.ict.superglu.util.StorageToken;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Map;

/**
 * A message class, for passing data between components.  Messages with special
 * meaning or messages intended to fit specific specifications should subclass
 * this message class.  This message that expands on the Tin Can API, in terms
 * of information available.
 *
 * @author auerbach
 */

public class Message extends BaseMessage {

    public static final String ID_KEY = "id";
    public static final String ACTOR_KEY = "actor";
    public static final String VERB_KEY = "verb";
    public static final String OBJECT_KEY = "object";
    public static final String RESULT_KEY = "result";
    public static final String SPEECH_ACT_KEY = "speechAct";
    public static final String TIMESTAMP_KEY = "timestamp";
    public static final String REQUEST_TEMPLATE = "requestTemplate";
    public static final String RESPONSE_TEMPLATE = "responseTemplate";


    //Context keys
    public static final String SESSION_ID_CONTEXT_KEY = "sessionId";
    public static final String CONTEXT_CONVERSATION_ID_KEY = "conversation-id";
    public static final String CONTEXT_IN_REPLY_TO_KEY = "in-reply-to";
    public static final String CONTEXT_REPLY_WITH_KEY = "reply-with";
    public static final String CONTEXT_REPLY_BY_KEY = "reply-by";
	public static final String PROPOSAL_KEY = "proposalId";
	public static final String CONTEXT_IN_REPLY_TO_MESSAGE = "reply-to-msg-id";


    public static DateFormat timestampFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS");

    /**
     * The actor performing the verb
     */
    private String actor;

    /**
     * A verb, which is an action that is occurring or has occurred
     */
    private String verb;

    /**
     * A target of the verb action, being acted upon or relating the actor and object
     */
    private String obj;

    /**
     * The outcome of this interaction
     */
    private Object result;

    /**
     * A speech act declaring the intent of this message, from SpeechActEnum
     */
    private SpeechActEnum speechAct;

    /**
     * The timestamp for when this message refers to, ISO 8601 formatted
     */
    private Date timestamp;
    
    /**
     * The Request Template is the template that we will be sending to the services expecting Response Template.
     */
    private MessageTemplate requestTemplate;
    
    /**
     * The Response Template is the template that we expect the services to provide us in response to Request Template.
     */
    private MessageTemplate responseTemplate;


    public Message(String actor, String verb, String obj, Object result, SpeechActEnum speechAct, Date timestamp, Map<String, Object> context, String id) {
        super(id, context);
        this.actor = actor;
        this.verb = verb;
        this.obj = obj;
        this.result = result;
        this.speechAct = speechAct;

        if (timestamp == null)
            this.timestamp = new Date();
        else
            this.timestamp = timestamp;
    }

    /**
     * 
     * Overriding Constructor to incorporate Request Template and Response Template.
     * 
     * @param actor
     * @param verb
     * @param obj
     * @param result
     * @param speechAct
     * @param timestamp
     * @param context
     * @param id
     * @param requestTemplate
     * @param responseTemplate
     */
    public Message(String actor, String verb, String obj, Object result, SpeechActEnum speechAct, Date timestamp, Map<String, Object> context, String id, MessageTemplate requestTemplate, MessageTemplate responseTemplate) {
        super(id, context);
        this.actor = actor;
        this.verb = verb;
        this.obj = obj;
        this.result = result;
        this.speechAct = speechAct;
        this.context = context;

        if (timestamp == null)
            this.timestamp = new Date();
        else
            this.timestamp = timestamp;
        
        if (requestTemplate != null)
        	this.requestTemplate = requestTemplate;
        
        if (responseTemplate != null)
        	this.responseTemplate = responseTemplate;
        
    }

    public Message() {
        super();
        this.actor = null;
        this.verb = null;
        this.obj = null;
        this.result = null;
        this.speechAct = SpeechActEnum.INFORM_ACT;
        this.timestamp = new Date();
        this.requestTemplate = null;
        this.responseTemplate = null;
    }


    //ACCESSORS
    public String getActor() {
        return actor;
    }


    public void setActor(String actor) {
        this.actor = actor;
    }


    public String getVerb() {
        return verb;
    }


    public void setVerb(String verb) {
        this.verb = verb;
    }


    public String getObj() {
        return obj;
    }


    public void setObj(String obj) {
        this.obj = obj;
    }


    public Object getResult() {
        return result;
    }


    public void setResult(Object result) {
        this.result = result;
    }


    public SpeechActEnum getSpeechAct() {
        return speechAct;
    }


    public void setSpeechAct(SpeechActEnum speechAct) {
        this.speechAct = speechAct;
    }


    public Date getTimestamp() {
        return timestamp;
    }


    public void setTimestamp(Date timestamp) {
        this.timestamp = timestamp;
    }


    public void updateTimestamp() {
        this.timestamp = new Date();
    }
    
    public MessageTemplate getRequestTemplate() {
		return requestTemplate;
	}

	public void setRequestTemplate(MessageTemplate requestTemplate) {
		this.requestTemplate = requestTemplate;
	}

	public MessageTemplate getResponseTemplate() {
		return responseTemplate;
	}

	public void setResponseTemplate(MessageTemplate responseTemplate) {
		this.responseTemplate = responseTemplate;
	}

	//Comparators
    @Override
    public boolean equals(Object otherObject) {
        if (!super.equals(otherObject))
            return false;
        if (!(otherObject instanceof Message))
            return false;

        Message other = (Message) otherObject;

        if (!fieldIsEqual(this.actor, other.actor))
            return false;
        if (!fieldIsEqual(this.verb, other.verb))
            return false;
        if (!fieldIsEqual(this.obj, other.obj))
            return false;
        if (!fieldIsEqual(this.result, other.result))
            return false;
        if (!fieldIsEqual(this.speechAct, other.speechAct))
            return false;
        if (!fieldIsEqual(this.timestamp, other.timestamp))
            return false;
        if (!fieldIsEqual(this.requestTemplate, other.requestTemplate))
            return false;
        if (!fieldIsEqual(this.responseTemplate, other.responseTemplate))
            return false;

        return true;
    }


    @Override
    /**
     *
     *   Generate a hash value for the message
     */
    public int hashCode() {
        int result = super.hashCode();
        int arbitraryPrimeNumber = 23;

        if (this.actor != null)
            result = result * arbitraryPrimeNumber + this.actor.hashCode();
        if (this.verb != null)
            result = result * arbitraryPrimeNumber + this.verb.hashCode();
        if (this.obj != null)
            result = result * arbitraryPrimeNumber + this.obj.hashCode();
        if (this.result != null)
            result = result * arbitraryPrimeNumber + this.result.hashCode();
        result = result * arbitraryPrimeNumber + this.speechAct.hashCode();
        if (this.timestamp != null)
            result = result * arbitraryPrimeNumber + this.timestamp.hashCode();
        if (this.requestTemplate != null)
            result = result * arbitraryPrimeNumber + this.requestTemplate.hashCode();
        if (this.responseTemplate != null)
            result = result * arbitraryPrimeNumber + this.responseTemplate.hashCode();

        return result;
    }


    //Serialization/Deserialization
    @Override
    public StorageToken saveToToken() {
        StorageToken result = super.saveToToken();

        result.setItem(ACTOR_KEY, this.actor);
        result.setItem(VERB_KEY, this.verb);
        result.setItem(OBJECT_KEY, this.obj);
        result.setItem(RESULT_KEY, SerializationConvenience.tokenizeObject(this.result));
        result.setItem(SPEECH_ACT_KEY, this.speechAct.toString());
        result.setItem(TIMESTAMP_KEY, timestampFormat.format(this.timestamp));

        return result;
    }


    @Override
    public void initializeFromToken(StorageToken token) {
        super.initializeFromToken(token);

        this.actor = (String) token.getItem(ACTOR_KEY, true, null);
        this.verb = (String) token.getItem(VERB_KEY, true, null);
        this.obj = (String) token.getItem(OBJECT_KEY, true, null);
        this.result = SerializationConvenience.untokenizeObject(token.getItem(RESULT_KEY, true, null));
        this.speechAct = SpeechActEnum.getEnum((String) token.getItem(SPEECH_ACT_KEY, true, SpeechActEnum.INFORM_ACT));
        try {
            this.timestamp = timestampFormat.parse((String) token.getItem(TIMESTAMP_KEY, true, null));
        } catch (Exception e) {
            e.printStackTrace();
            System.err.println("failed to parse timestamp of message, using current time as timestamp");
            this.timestamp = new Date();
        }
    }


}

File: javascript_module/src/core/message.js
'use strict'
const Zet = require('../util/zet'),
    Serialization = require('../util/serialization'),
    Serializable = Serialization.Serializable,
    tokenizeObject = Serialization.tokenizeObject,
    untokenizeObject = Serialization.untokenizeObject,
    Messaging = require('../core/messaging'),
    SUPERGLU_VERSION = require('../reference-data').version

const ACTOR_KEY = Messaging.ACTOR_KEY
    , VERB_KEY = Messaging.VERB_KEY
    , OBJECT_KEY = Messaging.OBJECT_KEY
    , RESULT_KEY = Messaging.RESULT_KEY
    , SPEECH_ACT_KEY = Messaging.SPEECH_ACT_KEY
    , TIMESTAMP_KEY = Messaging.TIMESTAMP_KEY
    , CONTEXT_KEY = Messaging.CONTEXT_KEY
    , PROPOSAL_KEY = Messaging.PROPOSAL_KEY
    , CONTEXT_CONVERSATION_ID_KEY = Messaging.CONTEXT_CONVERSATION_ID_KEY
    , CONTEXT_IN_REPLY_TO_KEY = Messaging.CONTEXT_IN_REPLY_TO_KEY
    , CONFIRM_PROPOSAl_ACT = Messaging.CONFIRM_PROPOSAl_ACT
    , PROPOSED_MESSAGE_ACT = Messaging.PROPOSED_MESSAGE_ACT
    , PROPOSED_MESSAGE_ACK_ACT = Messaging.PROPOSED_MESSAGE_ACK_ACT
    , REJECT_PROPOSAL_ACT = Messaging.REJECT_PROPOSAL_ACT

module.exports = Zet.declare({
    superclass: Serializable,
    CLASS_ID: 'Message',
    defineBody: function (self) {
        // Private Properties

        // Public Properties

        /** Create a Message
         @param actor: The actor who did or would do the given action
         @param verb: Some action that was or would be done by the actor
         @param obj: An object or target for the action
         @param result: The outcome of the action
         @param speechAct: A performative, stating why this message was sent
         @param context: A context object for the message, with additional data
         @param timestamp: A timestamp for when the message was created
         @param anId: A unique Id.  If none given, one will be assigned.
         **/
        self.construct = function construct(actor, verb, obj, result, speechAct,
                                            context, timestamp, anId) {
            self.inherited(construct, [anId])
            if (typeof actor === "undefined") {
                actor = null
            }
            if (typeof verb === "undefined") {
                verb = null
            }
            if (typeof obj === "undefined") {
                obj = null
            }
            if (typeof result === "undefined") {
                result = null
            }
            if (typeof speechAct === "undefined") {
                speechAct = Messaging.INFORM_ACT
            }
            if (typeof context === "undefined") {
                context = {}
            }
            if (typeof timestamp === "undefined") {
                timestamp = null
            }
            self._actor = actor
            self._verb = verb
            self._obj = obj
            self._result = result
            self._speechAct = speechAct
            self._timestamp = timestamp
            if (self._timestamp == null) {
                self.updateTimestamp()
            }
            // Fill in version keys
            if (!(Messaging.MESSAGE_VERSION_KEY in context)) {
                context[Messaging.MESSAGE_VERSION_KEY] = Messaging.VERSION
            }
            if (!(Messaging.SUPERGLU_VERSION_KEY in context)) {
                context[Messaging.SUPERGLU_VERSION_KEY] = SUPERGLU_VERSION
            }
            self._context = context
        }

        /** Get the actor for the message **/
        self.getActor = function getActor() {
            return self._actor
        }
        /** Set the actor for the message **/
        self.setActor = function setActor(value) {
            self._actor = value
        }

        /** Get the verb for the message **/
        self.getVerb = function getVerb() {
            return self._verb
        }
        /** Set the verb for the message **/
        self.setVerb = function setVerb(value) {
            self._verb = value
        }

        /** Get the object for the message **/
        self.getObject = function getObject() {
            return self._obj
        }
        /** Set the object for the message **/
        self.setObject = function setObject(value) {
            self._obj = value
        }

        /** Get the result for the message **/
        self.getResult = function getResult() {
            return self._result
        }
        /** Set the result for the message **/
        self.setResult = function setResult(value) {
            self._result = value
        }

        /** Get the speech act for the message **/
        self.getSpeechAct = function getSpeechAct() {
            return self._speechAct
        }
        /** Set the speech act for the message **/
        self.setSpeechAct = function setSpeechAct(value) {
            self._speechAct = value
        }

        /** Get the timestamp for the message (as an ISO-format string)**/
        self.getTimestamp = function getTimestamp() {
            return self._timestamp
        }
        /** Set the timestamp for the message (as an ISO-format string) **/
        self.setTimestamp = function setTimestamp(value) {
            self._timestamp = value
        }
        /** Update the timestamp to the current time **/
        self.updateTimestamp = function updateTimestamp() {
            self._timestamp = new Date().toISOString()
        }

        /** Check if the context field has a given key **/
        self.hasContextValue = function hasContextValue(key) {
            return (key in self._context) === true
        }

        /** Get all the keys for the context object **/
        self.getContextKeys = function getContextKeys() {
            var key, keys
            keys = []
            for (key in self._context) {
                keys.push(key)
            }
            return keys
        }

        /** Get the context value with the given key. If missing, return the default. **/
        self.getContextValue = function getContextValue(key, aDefault) {
            if (!(key in self._context)) {
                return aDefault
            }
            return self._context[key]
        }

        /** Set a context value with the given key-value pair **/
        self.setContextValue = function setContextValue(key, value) {
            self._context[key] = value
        }

        /** Delete the given key from the context **/
        self.delContextValue = function delContextValue(key) {
            delete self._context[key]
        }

        /** Save the message to a storage token **/
        self.saveToToken = function saveToToken() {
            var key, token, newContext, hadKey
            token = self.inherited(saveToToken)
            if (self._actor != null) {
                token.setitem(ACTOR_KEY, tokenizeObject(self._actor))
            }
            if (self._verb != null) {
                token.setitem(VERB_KEY, tokenizeObject(self._verb))
            }
            if (self._obj != null) {
                token.setitem(OBJECT_KEY, tokenizeObject(self._obj))
            }
            if (self._result != null) {
                token.setitem(RESULT_KEY, tokenizeObject(self._result))
            }
            if (self._speechAct != null) {
                token.setitem(SPEECH_ACT_KEY, tokenizeObject(self._speechAct))
            }
            if (self._timestamp != null) {
                token.setitem(TIMESTAMP_KEY, tokenizeObject(self._timestamp))
            }
            hadKey = false
            newContext = {}
            for (key in self._context) {
                hadKey = true
                newContext[tokenizeObject(key)] = tokenizeObject(self._context[key])
            }
            if (hadKey) {
                token.setitem(CONTEXT_KEY, tokenizeObject(newContext))
            }
            return token
        }

        /** Initialize the message from a storage token and some additional context (e.g., local objects) **/
        self.initializeFromToken = function initializeFromToken(token, context) {
            self.inherited(initializeFromToken, [token, context])
            self._actor = untokenizeObject(token.getitem(ACTOR_KEY, true, null), context)
            self._verb = untokenizeObject(token.getitem(VERB_KEY, true, null), context)
            self._obj = untokenizeObject(token.getitem(OBJECT_KEY, true, null), context)
            self._result = untokenizeObject(token.getitem(RESULT_KEY, true, null), context)
            self._speechAct = untokenizeObject(token.getitem(SPEECH_ACT_KEY, true, null), context)
            self._timestamp = untokenizeObject(token.getitem(TIMESTAMP_KEY, true, null), context)
            self._context = untokenizeObject(token.getitem(CONTEXT_KEY, true, {}), context)
        }
    }
})